File: scripts/bar.lua
local socket = require("socket")
local unix = socket.unix

local runtimeDir = os.getenv("XDG_RUNTIME_DIR")
local instanceSig = os.getenv("HYPRLAND_INSTANCE_SIGNATURE")

if not runtimeDir or not instanceSig then
  io.stderr:write("Error: XDG_RUNTIME_DIR or HYPRLAND_INSTANCE_SIGNATURE not found.\n")
  os.exit(1)
end

local socketPath = runtimeDir .. "/hypr/" .. instanceSig .. "/.socket2.sock"

local client, err = unix()
if not client then
  io.stderr:write("Failed to create unix socket: " .. err .. "\n")
  os.exit(1)
end

local ok, err = client:connect(socketPath)
if not ok then
  io.stderr:write("Failed to connect: " .. err .. "\n")
  os.exit(1)
end

client:send("subscribe:workspace\0")
client:send("subscribe:activewindow\0")
client:send("subscribe:activeworkspace\0")

local COMMANDS = {
  workspace = [[
    workspaces=$(
      hyprctl workspaces -j | jq -c '
        . as $ws
        | ($ws | map(.id)) as $existing_ids
        | [range(1;6)] as $all_ids
        | ($all_ids - $existing_ids) as $missing_ids
        | ($missing_ids | map({id: ., windows: 0})) as $missing_ws
        | ($ws + $missing_ws)
        | sort_by(.id)
      '
    );
    current_ws=$(hyprctl activeworkspace -j | jq -c .); 
    eww update WORKSPACES="$workspaces"; 
    eww update FOCUSED_WORKSPACE="$current_ws"
  ]],
  activewindow = [[
    active_win=$(hyprctl activewindow -j | jq -c .); 
    eww update ACTIVE_WINDOW="$active_win"
  ]],
}

local function runShellCommand(eventType)
  local cmd = COMMANDS[eventType]
  if not cmd then return end
  local handle = io.popen("/usr/bin/env bash -c '" .. cmd:gsub("'", "'\\''") .. "'")
  if handle then
    local result = handle:read("*a")
    handle:close()
    io.write(result)
  else
    io.stderr:write("Failed to execute command for " .. eventType .. "\n")
  end
end

local buffer = ""

while true do
  local data, err, partial = client:receive("*a")
  if not data and err ~= "timeout" then
    io.stderr:write("Socket error or closed: " .. tostring(err) .. "\n")
    break
  end
  buffer = buffer .. (data or partial or "")
  -- Process lines in buffer
  while true do
    local lineEnd = buffer:find("\n", 1, true)
    if not lineEnd then break end
    local line = buffer:sub(1, lineEnd - 1)
    buffer = buffer:sub(lineEnd + 1)
    for eventType, _ in pairs(COMMANDS) do
      if line:find(eventType) then
        runShellCommand(eventType)
        break
      end
    end
  end
  socket.sleep(0.01) -- avoid busy loop
end

client:close()

File: scripts/bar.js
import { createConnection } from "net";
import { exec } from "child_process";

const runtimeDir = process.env.XDG_RUNTIME_DIR;
const instanceSig = process.env.HYPRLAND_INSTANCE_SIGNATURE;

if (!runtimeDir || !instanceSig) {
  console.error(
    "Error: environment variables XDG_RUNTIME_DIR or HYPRLAND_INSTANCE_SIGNATURE not found."
  );
  process.exit(1);
}

const socketPath = `${runtimeDir}/hypr/${instanceSig}/.socket2.sock`;

const client = createConnection(socketPath, () => {
  console.log("Connected to Hyprland IPC socket, subscribing to events...");
  client.write("subscribe:workspace\0");
  client.write("subscribe:activewindow\0");
  client.write("subscribe:activeworkspace\0");

  for (const event of Object.keys(COMMANDS)) {
    runShellCommand(event);
  }
});

client.on("error", (err) => {
  console.error(`Failed to connect to Hyprland IPC socket: ${err.message}`);
  process.exit(1);
});

const COMMANDS = {
  workspace: `
    workspaces=$(
      hyprctl workspaces -j | jq -c '
        . as $ws
        | ($ws | map(.id)) as $existing_ids
        | [range(1;6)] as $all_ids
        | ($all_ids - $existing_ids) as $missing_ids
        | ($missing_ids | map({id: ., windows: 0})) as $missing_ws
        | ($ws + $missing_ws)
        | sort_by(.id)
        '

    );
    current_ws=$(hyprctl activeworkspace -j | jq -c .); 
    eww update WORKSPACES="$workspaces"; 
    eww update FOCUSED_WORKSPACE="$current_ws"
  `,
  activewindow: `
    active_win=$(hyprctl activewindow -j | jq -c .); 
    eww update ACTIVE_WINDOW="$active_win"
  `,
};

function runShellCommand(eventType) {
  const cmd = COMMANDS[eventType];
  if (!cmd) return;

  exec(
    `/usr/bin/env bash -c '${cmd.replace(/'/g, `'\\''`)}'`,
    { timeout: 10000 },
    (error, stdout, stderr) => {
      if (error) {
        console.error(
          `Subprocess error for ${eventType}: ${stderr || error.message}`
        );
      } else {
        process.stdout.write(stdout);
      }
    }
  );
}

let buffer = "";

client.on("data", (data) => {
  buffer += data.toString();

  let index;
  while ((index = buffer.indexOf("\n")) >= 0) {
    const message = buffer.slice(0, index);
    buffer = buffer.slice(index + 1);

    for (const eventType of Object.keys(COMMANDS)) {
      if (message.includes(eventType)) {
        runShellCommand(eventType);
        break;
      }
    }
  }
});

client.on("end", () => {
  console.log("Disconnected from Hyprland IPC.");
});

process.on("SIGINT", () => {
  console.log("Exiting...");
  client.end();
  process.exit();
});
